source/exe/main_common.cc

```diff
@@ -93,16 +93,21 @@ bool MainCommonBase::run() {
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void MainCommonBase::adminRequest(absl::string_view path_and_query, absl::string_view method,
-                                  const AdminRequestFn& handler) {
+std::future<AdminResponse> MainCommonBase::adminRequest(absl::string_view path_and_query,
+                                                        absl::string_view method) {
   std::string path_and_query_buf = std::string(path_and_query);
   std::string method_buf = std::string(method);
-  server_->dispatcher().post([this, path_and_query_buf, method_buf, handler]() {
-    Http::HeaderMapImpl response_headers;
-    std::string body;
-    server_->admin().request(path_and_query_buf, method_buf, response_headers, body);
-    handler(response_headers, body);
-  });
+  // Note: must be a shared_ptr to allow the std::function to be copyable.
+  auto response_promise = std::make_shared<std::promise<AdminResponse>>();
+  std::future<AdminResponse> response_future = response_promise->get_future();
+  server_->dispatcher().post(
+      [ this, path_and_query_buf, method_buf, response_promise{std::move(response_promise)} ]() {
+        AdminResponse response;
+        response.headers = std::make_unique<Http::HeaderMapImpl>();
+        server_->admin().request(path_and_query_buf, method_buf, *response.headers, response.body);
+        response_promise->set_value(std::move(response));
```

Comment:
They do provide some sort of internal locking.  A promise and future (created from that promise) act like a single use write-read channel.  As long as the promise is written to once and the future is read from on a single thread, it is guaranteed to be safe (other types of accesses like reading from the future on multiple threads, etc, are not guaranteed to be thread safe).  The most common use is the return value from `std::async`, which is basically a dispatcher `post()` with a return value.

Reply:
OK. This seems fine to me if we want to move in this direction, especially for code that isn't really performance critical.

Reply:
Right, I think this sort of thing really only makes sense at the boundary since we can't assume that the caller is running within a dispatcher thread.